README
NAME
    Jifty::Plugin::OpenID - Provides OpenID authentication for your jifty
    app

DESCRIPTION
    Provides OpenID authentication for your app

USAGE
  Config
    please provide "OpenIDSecret" in your etc/config.yml , the "OpenIDUA" is
    optional , OpenID Plugin will use LWPx::ParanoidAgent by default.

        --- 
        application:
            OpenIDSecret: 1234
            OpenIDUA: LWP::UserAgent

    or you can set "OpenIDUserAgent" environment var in command-line:

        OpenIDUserAgent=LWP::UserAgent bin/jifty server

    if you are using LWPx::ParanoidAgent as your openid agent. you will need
    to provide "JIFTY_OPENID_WHITELIST_HOST" for your own OpenID server.

        export JIFTY_OPENID_WHITELIST_HOST=123.123.123.123

  User Model
    Create your user model , and let it uses
    Jifty::Plugin::OpenID::Mixin::Model::User to mixin "openid" column. and
    a "name" method.

        use TestApp::Record schema {

            column email =>
                type is 'varchar';

        };
        use Jifty::Plugin::OpenID::Mixin::Model::User;

        sub name {
            my $self = shift;
            return $self->email;
        }

    Note: you might need to declare a "name" method. because the OpenID
    CreateOpenIDUser action and SkeletonApp needs current_user->username to
    show welcome message and success message , which calls
    "brief_description" method. See Jifty::Record for "brief_description"
    method.

  View
    OpenID plugin provides AuthenticateOpenID Action. so that you can render
    an AuthenticateOpenID in your template:

        form {
            my $openid = new_action( class   => 'AuthenticateOpenID',
                                    moniker => 'authenticateopenid' );
            render_action( $openid );
        };

    this action renders a form which provides openid url field. and you will
    need to provide a submit button in your form.

        form {
            my $openid = new_action( class   => 'AuthenticateOpenID',
                                    moniker => 'authenticateopenid' );

            # ....

            render_action( $openid );
            outs_raw(
                Jifty->web->return(
                    to     => '/openid_verify_done',
                    label  => _("Login with OpenID"),
                    submit => $openid
                ));
        };

    the "to" field is for verified user to redirect to. so that you will
    need to implement a template called "/openid_verify_done":

        template '/openid_verify_done' => page {
            h1 { "Done" };
        };

AUTHORS
    Alex Vandiver, Cornelius <cornelius.howl {at} gmail.com >

LICENSE
    Copyright 2005-2009 Best Practical Solutions, LLC.

    This program is free software and may be modified and distributed under
    the same terms as Perl itself.



Changes
Revision history for Perl module Jifty::Plugin::OpenID

1.00 Wed, 10 Jun 2009 11:55:15 +0200
    - bump version number to avoid conflict with debian
    - add minimal load test
    - temporary suspend tests with Test::OpenID::Server

0.10 Tue, 09 Jun 2009 22:28:10 +0200
    - original version for CPAN


t/TestApp/lib/TestApp/View.pm
package TestApp::View;
use strict;
use warnings;
use Jifty::View::Declare -base;
use Jifty::View::Declare::Helpers;


template '/' => page {
    with ( name => 'openid-form' ),
    form {
        my $openid = new_action( class   => 'AuthenticateOpenID',
                                moniker => 'authenticateopenid' );
        div { { class is 'openid'};
            div { { id is 'openid-login' };
                render_action( $openid );
                outs_raw(
                    Jifty->web->return(
                        to     => '/openid_verify_done',
                        label  => _("Login with OpenID"),
                        submit => $openid
                    ));
            };
        };
    };
};


template '/openid_verify_done' => page {
    h1 { "Done" };
};




1;


t/TestApp/lib/TestApp/Model/User.pm
use strict;
use warnings;

package TestApp::Model::User;
use Jifty::DBI::Schema;

use TestApp::Record schema {

column email =>
   type is 'varchar';

};

use Jifty::Plugin::OpenID::Mixin::Model::User;

1;



t/TestApp/lib/TestApp/Action/User.pm
use strict;
use warnings;

=head1 NAME

TestApp::Action::User

=cut

package TestApp::Action::User;
use base qw/TestApp::Action Jifty::Action/;

use Jifty::Param::Schema;
use Jifty::Action schema {

};

=head2 take_action

=cut

sub take_action {
    my $self = shift;
    
    # Custom action code
    
    $self->report_success if not $self->result->failure;
    
    return 1;
}

=head2 report_success

=cut

sub report_success {
    my $self = shift;
    # Your success message here
    $self->result->message('Success');
}

1;



lib/Jifty/Plugin/OpenID/View.pm
package Jifty::Plugin::OpenID::View;
use strict;
use warnings;
use Jifty::View::Declare -base;

=head1 NAME

Jifty::Plugin::OpenID::View - Login and create pages for OpenID plugin

=head1 DESCRIPTION

The view class for L<Jifty::Plugin::OpenID>.  Provides login and create pages.

=cut

template 'openid/login' => page {
    { title is _("Login with your OpenID") }
    my ($action, $next) = get('action', 'next');

    div {
        unless ( Jifty->web->current_user->id ) {
            div {
                attr { id => 'openid-login' };
                outs(
                    p {
                        em {
                            _(  qq{If you have a Livejournal or other OpenID account, you don\'t even need to sign up. Just log in.}
                            );
                        }
                    }
                );
                Jifty->web->form->start( call => $next );
                render_action($action);
                form_submit(
                    label  => _("Go for it!"),
                    submit => $action
                );
                Jifty->web->form->end;
            };
        }
        else {
            outs( _("You already logged in.") );
        }
    }
};

template 'openid/create' => page {
    title is 'Set your username';
    my ( $action, $next ) = get( 'action', 'next' );

    p {
        outs(
            _(  'We need you to set a username or quickly check the one associated with your OpenID. Your username is what other people will see when you ask questions or make suggestions'
            )
        );
    };
    p {
        outs(
            _(  'If the username provided conflicts with an existing username or contains invalid characters, you will have to give us a new one.'
            )
        );
    };
    Jifty->web->form->start( call => $next , name => 'openid-user-create' );
    render_action($action);
    form_submit( label => _('Continue'), submit => $action );
    Jifty->web->form->end;
};


1;


lib/Jifty/Plugin/OpenID/Dispatcher.pm
use strict;
use warnings;

package Jifty::Plugin::OpenID::Dispatcher;
use Jifty::Dispatcher -base;

=head1 NAME

Jifty::Plugin::OpenID::Dispatcher - Dispatcher for OpenID plugin

=head1 DESCRIPTION

Dispatcher for L<Jifty::Plugin::OpenID>.  Handles a lot of the work.

=cut

before qr'^/(?:openid/link)' => run {
    tangent('/openid/login') unless (Jifty->web->current_user->id)
};

before qr'^/openid/login' => run {
    Jifty->api->allow('AuthenticateOpenID');
    set action => Jifty->web->new_action(
        class   => 'AuthenticateOpenID',
        moniker => 'authenticateopenid'
    );
};

before qr'^/openid/verify' => run {
    Jifty->api->allow('VerifyOpenID');
    Jifty->web->request->add_action(
        class   => 'VerifyOpenID',
        moniker => 'verifyopenid'
    );
};

on 'openid/verify_and_link' => run {
    my $result = Jifty->web->response->result('verifyopenid');
    my $user   = Jifty->web->current_user;
    if ( defined $result and $result->success and $user->id ) {
        my $openid = $result->content('openid');
        my ( $ret, $msg ) = $user->user_object->validate_openid( $openid );

        if ( not $ret ) {
            $result->error(_("It looks like someone is already using that OpenID."));
            redirect '/openid/link';
        }
        else {
            $user->user_object->link_to_openid( $openid );
            $result->message(_("The OpenID '$openid' has been linked to your account."));
        }
    }
    redirect '/';
};

on 'openid/verify_and_login' => run {
    my $result = Jifty->web->response->result('verifyopenid');

    if ( defined $result and $result->success ) {
        my $openid = $result->content('openid');
        my $user = Jifty->app_class('CurrentUser')->new( openid => $openid );
        Jifty->log->info("User Class: $user. OpenID: $openid");

        if ( $user->id ) {
            # Set up our login message
            $result->message( _("Welcome back, ") . $user->username . "." );

            # Actually do the signin thing.
            Jifty->web->current_user($user);
            Jifty->web->session->expires( undef );
            Jifty->web->session->set_cookie;

            if(Jifty->web->request->continuation) {
                Jifty->web->request->continuation->call;
            } else {
                redirect '/';
            }
        }
        else {
            # User needs to create account still
            Jifty->web->session->set( openid => $openid );
            Jifty->log->info("got openid: $openid");
            my $nick = get('openid.sreg.nickname');
            if ( $nick ) {
                redirect( Jifty::Web::Form::Clickable->new( url => '/openid/create', parameters => { 
                    nickname => $nick, 
                    openid => $openid } ));
            }
            else {
                redirect( Jifty::Web::Form::Clickable->new( url => '/openid/create' ) );
            }
        }
    }
    else {
        if(Jifty->web->request->continuation) {
            Jifty->web->request->continuation->call;
        } else {
            redirect '/openid/login';
        }
    }
};

on 'openid/create' => run {
    if ( not Jifty->web->session->get('openid') ) {
        redirect '/';
    }

    set action => Jifty->web->new_action( class => 'CreateOpenIDUser', parameters => { openid => Jifty->web->session->get("openid") } );
    set 'next' => Jifty->web->request->continuation ||
                  Jifty::Continuation->new( request => Jifty::Request->new( path => "/" ) );
};

1;


lib/Jifty/Plugin/OpenID.pm
use strict;
use warnings;

package Jifty::Plugin::OpenID;
use base qw/Jifty::Plugin/;

our $VERSION = '1.00';

=head1 NAME

Jifty::Plugin::OpenID - Provides OpenID authentication for your jifty app

=head1 DESCRIPTION

Provides OpenID authentication for your app

=head1 USAGE

=head2 Config

please provide C<OpenIDSecret> in your F<etc/config.yml> , the C<OpenIDUA> is
B<optional> , OpenID Plugin will use L<LWPx::ParanoidAgent> by default.

    --- 
    application:
        OpenIDSecret: 1234
        OpenIDUA: LWP::UserAgent

or you can set C<OpenIDUserAgent> environment var in command-line:

    OpenIDUserAgent=LWP::UserAgent bin/jifty server

if you are using L<LWPx::ParanoidAgent> as your openid agent. 
you will need to provide C<JIFTY_OPENID_WHITELIST_HOST> for your own OpenID server.

    export JIFTY_OPENID_WHITELIST_HOST=123.123.123.123

=head2 User Model

Create your user model , and let it uses
L<Jifty::Plugin::OpenID::Mixin::Model::User> to mixin "openid" column.
and a C<name> method.

    use TestApp::Record schema {

        column email =>
            type is 'varchar';

    };
    use Jifty::Plugin::OpenID::Mixin::Model::User;

    sub name {
        my $self = shift;
        return $self->email;
    }

Note: you might need to declare a C<name> method. because the OpenID
CreateOpenIDUser action and SkeletonApp needs current_user->username to show
welcome message and success message , which calls C<brief_description> method.
See L<Jifty::Record> for C<brief_description> method.

=head2 View

OpenID plugin provides AuthenticateOpenID Action. so that you can render an
AuthenticateOpenID in your template:

    form {
        my $openid = new_action( class   => 'AuthenticateOpenID',
                                moniker => 'authenticateopenid' );
        render_action( $openid );
    };

this action renders a form which provides openid url field.
and you will need to provide a submit button in your form.  

    form {
        my $openid = new_action( class   => 'AuthenticateOpenID',
                                moniker => 'authenticateopenid' );

        # ....

        render_action( $openid );
        outs_raw(
            Jifty->web->return(
                to     => '/openid_verify_done',
                label  => _("Login with OpenID"),
                submit => $openid
            ));
    };

the C<to> field is for verified user to redirect to.
so that you will need to implement a template called C</openid_verify_done>:

    template '/openid_verify_done' => page {
        h1 { "Done" };
    };

=cut

sub init {
    my $self = shift;
    my %opt = @_;
    my $ua_class = $self->get_ua_class;
    eval "require $ua_class";
}

sub get_ua_class {
    return Jifty->config->app('OpenIDUA') 
                || $ENV{OpenIDUserAgent} 
                || 'LWPx::ParanoidAgent' ;
}

sub new_ua {
    my $class = shift;
    my $ua;
    my $ua_class = $class->get_ua_class;

    Jifty->log->info( "OpenID Plugin is using $ua_class as UserAgent" );

    if( $ua_class eq 'LWPx::ParanoidAgent' ) {
         $ua = LWPx::ParanoidAgent->new(
                        whitelisted_hosts => [ $ENV{JIFTY_OPENID_WHITELIST_HOST} ]
                     );
    }
    else {
        $ua = $ua_class->new;
    }
    return $ua;
}


sub get_csr {
    my $class = shift;
    return Net::OpenID::Consumer->new(
        ua              => $class->new_ua ,
        cache           => Cache::FileCache->new,
        args            => scalar Jifty->handler->cgi->Vars,
        consumer_secret => Jifty->config->app('OpenIDSecret'),
        @_,
    );
}

=head1 AUTHORS

Alex Vandiver, Cornelius  <cornelius.howl {at} gmail.com >

=head1 LICENSE

Copyright 2005-2009 Best Practical Solutions, LLC.

This program is free software and may be modified and distributed under the same terms as Perl itself.

=cut

1;
